File: commit.swift
#!/usr/bin/env swift

import Foundation

// 讀取 `.env` 並載入環境變數
func loadEnv() {
    let envPath = ".env"
    guard let envContent = try? String(contentsOfFile: envPath, encoding: .utf8) else {
        print("⚠️ 未找到 .env 檔案，請確認檔案是否存在")
        return
    }

    for line in envContent.split(separator: "\n") {
        let parts = line.split(separator: "=", maxSplits: 1)
        if parts.count == 2 {
            let key = parts[0].trimmingCharacters(in: .whitespaces)
            let value = parts[1].trimmingCharacters(in: .whitespaces)
            setenv(key, value, 1) // 設置環境變數
        }
    }
}

// 先讀取 `.env`
loadEnv()

// 獲取 API Key
let apiKey = ProcessInfo.processInfo.environment["OPEN_API_KEY_COMMIT"] ?? ""

if apiKey.isEmpty {
    print("❌ ERROR: 未設置 OPEN_API_KEY_COMMIT")
    exit(1)
}

// 執行 Shell 指令並獲取輸出
func runShellCommand(_ command: String) -> String {
    let process = Process()
    let pipe = Pipe()
    
    process.launchPath = "/bin/bash"
    process.arguments = ["-c", command]
    process.standardOutput = pipe
    process.standardError = pipe
    
    let fileHandle = pipe.fileHandleForReading
    process.launch()
    
    let outputData = fileHandle.readDataToEndOfFile()
    return String(data: outputData, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
}

// 取得 Git 差異內容
func getGitDiff() -> String {
    let changedFiles = runShellCommand("git diff --cached --name-only")
        .split(separator: "\n")
        .filter { !$0.isEmpty && !$0.contains(".min.") }
    
    if changedFiles.isEmpty {
        print("⚠️ No changes to commit.")
        return ""
    }

    print("📂 變更的檔案: \(changedFiles)")
    
    var diffText = ""
    for file in changedFiles {
        let diff = runShellCommand("git diff --cached -- \"\(file)\"")
        diffText += diff + "\n"
    }

    return String(diffText.prefix(4000)) // 限制 4000 字符
}

// 透過 OpenAI 生成 Commit 訊息
func generateCommitMessage(diff: String) -> String {
    let apiUrl = "https://api.chatanywhere.org/v1/chat/completions" // 修改為你自己的 API URL
    guard let url = URL(string: apiUrl) else { return "Refactor code." }
    
    let requestData: [String: Any] = [
        "model": "gpt-4o-mini",
        "messages": [
            ["role": "system", "content": "你是一個優秀的開發者，負責撰寫簡潔又描述清楚的 Git commit 訊息。"],
            ["role": "user", "content": "根據以下的 git 差異生成一個有意義的 commit 英文動詞開頭加繁體中文訊息，請包含兩個部分：\n1. 範例: [New|Update|Remove|Refactor|Fix|Misc]（不超過80字）\n\n Description:（條列描述變更內容）：\n\(diff)"]
        ]
    ]
    
    var request = URLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    
    do {
        request.httpBody = try JSONSerialization.data(withJSONObject: requestData, options: [])
    } catch {
        print("❌ JSON 序列化錯誤: \(error)")
        return "Refactor code."
    }
    
    let semaphore = DispatchSemaphore(value: 0)
    var commitMessage = "Refactor code."

    URLSession.shared.dataTask(with: request) { data, response, error in
        if let error = error {
            print("❌ API 錯誤: \(error)")
        } else if let data = data, 
                  let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any],
                  let choices = jsonResponse["choices"] as? [[String: Any]],
                  let message = choices.first?["message"] as? [String: Any],
                  let content = message["content"] as? String {
            commitMessage = content.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        semaphore.signal()
    }.resume()
    
    semaphore.wait()
    return commitMessage
}

// 🏁 主執行流程
let diff = getGitDiff()
if !diff.isEmpty {
    let commitMessage = generateCommitMessage(diff: diff)
    print("✨ Commit 訊息: \n\(commitMessage)")
}

File: commit.py
import os
import subprocess
import requests
from dotenv import load_dotenv

# 加載 .env 環境變數
load_dotenv()

# 使用 OpenAI API Key
API_KEY = os.getenv("OPEN_API_KEY_COMMIT")
API_URL = "https://api.chatanywhere.org/v1/chat/completions"  # 要改成自己用的 URL

# 取得 Git 差異內容
def get_git_diff():
    try:
        # 獲取所有已 staged 的檔案
        changed_files = subprocess.run(
            ["git", "diff", "--cached", "--name-only"],
            capture_output=True,
            text=True,
            check=True
        ).stdout.strip().split("\n")

        changed_files = [file for file in changed_files if file and ".min." not in file]

        if not changed_files:
            print("No files changed.")
            return ""

        print("判斷檔案：", changed_files)

        # 取得這些檔案的 diff 內容
        diffs = []
        for file in changed_files:
            try:
                diff_output = subprocess.run(
                    ["git", "diff", "--cached", "--", file],
                    capture_output=True,
                    text=True,
                    check=True
                ).stdout
                diffs.append(diff_output)
            except subprocess.CalledProcessError as e:
                print(f"Error getting diff for file {file}: {e}")
                continue

        diff_text = "\n".join(diffs)

        # 限制 diff 內容不超過 4000 字符
        return diff_text[:4000]  # 預留部分空間給 API 使用
    except subprocess.CalledProcessError as e:
        print("Error fetching git diff:", e)
        return ""

# 使用 OpenAI API 生成 commit 訊息
def generate_commit_message(diff_text):
    if not API_KEY:
        print("Error: OPEN_API_KEY_COMMIT is not set in .env file.")
        return "Refactor code."

    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json"
    }
    payload = {
        "model": "gpt-4o-mini",  # 使用 GPT-4 模型
        "messages": [
            {"role": "system", "content": "你是一個優秀的開發者，負責撰寫簡潔又描述清楚的 Git commit 訊息, 和建議並檢查變數命名是否有swift規範(小駝峰命名法:第1個字為小寫駝峰)。"},
            {"role": "user", "content": f"根據以下的 git 差異生成一個有意義的 commit 英文動詞開頭加繁體中文訊息和建議，請包含3部分：\n 範例: [New|Update|Remove|Refactor|Fix|Misc]（不超過80字）\n\n Description:（條列描述變更內容）：\n{diff_text}\n\n Suggest :\n [⚠️) 駝峰檢查], 檔案:L行 ==> 變數"}
        ]
    }

    try:
        response = requests.post(API_URL, json=payload, headers=headers)
        response.raise_for_status()  # 如果發生 HTTP 錯誤會丟出異常:
        data = response.json()

        commit_message = data.get("choices", [{}])[0].get("message", {}).get("content", "").strip()
        return commit_message if commit_message else "Refactor code."
    except requests.exceptions.RequestException as e:
        print("Error generating commit message:", e)
        return "Refactor code."

# 主執行流程
def main():
    diff = get_git_diff()
    if not diff:
        print("No changes to commit.")
        return

    commit_message = generate_commit_message(diff)
    print(commit_message)

if __name__ == "__main__":
    main()

